File: java-bigdecimal2/src/main/java/src/Bench.java
package src;

import org.openjdk.jmh.annotations.*;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

@BenchmarkMode(Mode.SingleShotTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Fork(value = 1, jvmArgs = {"-Xms4g", "-Xmx4g"})
public class Bench {
    public void decimalOp(DecimalOp op, BigDecimal x, BigDecimal y) {
        switch (op) {
            case ADD: x.add(y); break;
            case SUB: x.subtract(y); break;
            case MUL: x.multiply(y); break;
            case DIV: x.divide(y, 2, RoundingMode.CEILING);break;
            default:
        }
    }

    public void foldOp(DecimalOp op, BigDecimal init, int count) {
        List<BigDecimal> values = IntStream.rangeClosed(1, count)
                .mapToObj(i -> new BigDecimal(i))
                .collect(Collectors.toList());

        BigDecimal acc = init;

        switch (op) {
            case ADD:
                for (BigDecimal v : values) acc = acc.add(v);
                break;
            case SUB:
                for (BigDecimal v : values) acc = acc.subtract(v);
                break;
            case MUL:
                for (BigDecimal v : values) acc = acc.multiply(v);
                break;
            case DIV:
                for (BigDecimal v : values) acc = acc.divide(v, 2 , RoundingMode.CEILING);break;
            default:
        }
    }

//    @TearDown
//    public void tearDown() {
//        System.gc();
//    }
    /* ADD */
    @Benchmark
    public void addSelf() {decimalOp(DecimalOp.ADD, new BigDecimal("2.01"), new BigDecimal("2.01"));}
    @Benchmark
    public void addSimple() {decimalOp(DecimalOp.ADD, new BigDecimal("2"), new BigDecimal("1"));}
    @Benchmark
    public void addOne() {decimalOp(DecimalOp.ADD, new BigDecimal("2.01"), new BigDecimal("1"));}
    @Benchmark
    public void addTwo() {decimalOp(DecimalOp.ADD, new BigDecimal("2.01"), new BigDecimal("2"));}
    @Benchmark
    public void addOneHundred() {decimalOp(DecimalOp.ADD, new BigDecimal("2.01"), new BigDecimal("100"));}
    @Benchmark
    public void addPointZeroOne() {decimalOp(DecimalOp.ADD, new BigDecimal("2.01"), new BigDecimal("0.01"));}
    @Benchmark
    public void addNegativePointFive() {decimalOp(DecimalOp.ADD, new BigDecimal("2.01"), new BigDecimal("-0.5"));}
    @Benchmark
    public void addPi() {decimalOp(DecimalOp.ADD, new BigDecimal("2.01"), new BigDecimal("3.1415926535897932384626433832"));}
    @Benchmark
    public void addNegativePi() {decimalOp(DecimalOp.ADD, new BigDecimal("2.01"), new BigDecimal("-3.1415926535897932384626433832"));}
    @Benchmark
    public void add10k() {foldOp(DecimalOp.ADD, new BigDecimal("0"), 10000);}

    /* SUB */
    @Benchmark
    public void subSelf() {decimalOp(DecimalOp.SUB, new BigDecimal("2.01"), new BigDecimal("2.01"));}
    @Benchmark
    public void subSimple() {decimalOp(DecimalOp.SUB, new BigDecimal("2"), new BigDecimal("1"));}
    @Benchmark
    public void subOne() {decimalOp(DecimalOp.SUB, new BigDecimal("2.01"), new BigDecimal("1"));}
    @Benchmark
    public void subTwo() {decimalOp(DecimalOp.SUB, new BigDecimal("2.01"), new BigDecimal("2"));}
    @Benchmark
    public void subOneHundred() {decimalOp(DecimalOp.SUB, new BigDecimal("2.01"), new BigDecimal("100"));}
    @Benchmark
    public void subPointZeroOne() {decimalOp(DecimalOp.SUB, new BigDecimal("2.01"), new BigDecimal("0.01"));}
    @Benchmark
    public void subNegativePointFive() {decimalOp(DecimalOp.SUB, new BigDecimal("2.01"), new BigDecimal("-0.5"));}
    @Benchmark
    public void subPi() {decimalOp(DecimalOp.SUB, new BigDecimal("2.01"), new BigDecimal("3.1415926535897932384626433832"));}
    @Benchmark
    public void subNegativePi() {decimalOp(DecimalOp.SUB, new BigDecimal("2.01"), new BigDecimal("-3.1415926535897932384626433832"));}
    @Benchmark
    public void sub10k() {foldOp(DecimalOp.SUB, new BigDecimal("5000000"), 10000);}


    /* MUL */
    @Benchmark
    public void mulSelf() {decimalOp(DecimalOp.MUL, new BigDecimal("2.01"), new BigDecimal("2.01"));}
    @Benchmark
    public void mulSimple() {decimalOp(DecimalOp.MUL, new BigDecimal("2"), new BigDecimal("1"));}
    @Benchmark
    public void mulOne() {decimalOp(DecimalOp.MUL, new BigDecimal("2.01"), new BigDecimal("1"));}
    @Benchmark
    public void mulTwo() {decimalOp(DecimalOp.MUL, new BigDecimal("2.01"), new BigDecimal("2"));}
    @Benchmark
    public void mulOneHundred() {decimalOp(DecimalOp.MUL, new BigDecimal("2.01"), new BigDecimal("100"));}
    @Benchmark
    public void mulPointZeroOne() {decimalOp(DecimalOp.MUL, new BigDecimal("2.01"), new BigDecimal("0.01"));}
    @Benchmark
    public void mulNegativePointFive() {decimalOp(DecimalOp.MUL, new BigDecimal("2.01"), new BigDecimal("-0.5"));}
    @Benchmark
    public void mulPi() {decimalOp(DecimalOp.MUL, new BigDecimal("2.01"), new BigDecimal("3.1415926535897932384626433832"));}
    @Benchmark
    public void mulNegativePi() {decimalOp(DecimalOp.MUL, new BigDecimal("2.01"), new BigDecimal("-3.1415926535897932384626433832"));}
    @Benchmark
    public void mul10k() {foldOp(DecimalOp.MUL, new BigDecimal("1.1"), 25);}

    /* DIV */
    @Benchmark
    public void divSelf() {decimalOp(DecimalOp.DIV, new BigDecimal("2.01"), new BigDecimal("2.01"));}
    @Benchmark
    public void divSimple() {decimalOp(DecimalOp.DIV, new BigDecimal("2"), new BigDecimal("1"));}
    @Benchmark
    public void divOne() {decimalOp(DecimalOp.DIV, new BigDecimal("2.01"), new BigDecimal("1"));}
    @Benchmark
    public void divTwo() {decimalOp(DecimalOp.DIV, new BigDecimal("2.01"), new BigDecimal("2"));}
    @Benchmark
    public void divOneHundred() {decimalOp(DecimalOp.DIV, new BigDecimal("2.01"), new BigDecimal("100"));}
    @Benchmark
    public void divPointZeroOne() {decimalOp(DecimalOp.DIV, new BigDecimal("2.01"), new BigDecimal("0.01"));}
    @Benchmark
    public void divNegativePointFive() {decimalOp(DecimalOp.DIV, new BigDecimal("2.01"), new BigDecimal("-0.5"));}
    @Benchmark
    public void divPi() {decimalOp(DecimalOp.DIV, new BigDecimal("2.01"), new BigDecimal("3.1415926535897932384626433832"));}
    @Benchmark
    public void divNegativePi() {decimalOp(DecimalOp.DIV, new BigDecimal("2.01"), new BigDecimal("-3.1415926535897932384626433832"));}
    @Benchmark
    public void div10k() {foldOp(DecimalOp.DIV, new BigDecimal("1000000000"), 10000);}
}

File: rust_decimal/benches/bench.rs
#![feature(test)]

extern crate test;

use core::str::FromStr;
use rust_decimal::Decimal;
use test::Bencher;

macro_rules! bench_decimal_op {
    ($name:ident, $op:tt, $x:expr, $y:expr) => {
        #[bench]
        fn $name(b: &mut Bencher) {
            let x = Decimal::from_str($x).unwrap();
            let y = Decimal::from_str($y).unwrap();

            b.iter(|| {
                let result = x $op y;
                ::test::black_box(result);
            });
        }
    }
}


macro_rules! bench_fold_op {
    ($name:ident, $op:tt, $init:expr, $count:expr) => {
        #[bench]
        fn $name(b: &mut ::test::Bencher) {
            fn fold(values: &[Decimal]) -> Decimal {
                let mut acc: Decimal = $init.into();
                for value in values {
                    acc = acc $op value;
                }
                acc
            }

            let values: Vec<Decimal> = test::black_box((1..$count).map(|i| i.into()).collect());
            b.iter(|| {
                let result = fold(&values);
                ::test::black_box(result);
            });
        }
    }
}

/* Add */
bench_decimal_op!(add_self, +, "2.01", "2.01");
bench_decimal_op!(add_simple, +, "2", "1");
bench_decimal_op!(add_one, +, "2.01", "1");
bench_decimal_op!(add_two, +, "2.01", "2");
bench_decimal_op!(add_one_hundred, +, "2.01", "100");
bench_decimal_op!(add_point_zero_one, +, "2.01", "0.01");
bench_decimal_op!(add_negative_point_five, +, "2.01", "-0.5");
bench_decimal_op!(add_pi, +, "2.01", "3.1415926535897932384626433832");
bench_decimal_op!(add_negative_pi, +, "2.01", "-3.1415926535897932384626433832");

bench_fold_op!(add_10k, +, 0, 10_000);

/* Sub */
bench_decimal_op!(sub_self, -, "2.01", "2.01");
bench_decimal_op!(sub_simple, -, "2", "1");
bench_decimal_op!(sub_one, -, "2.01", "1");
bench_decimal_op!(sub_two, -, "2.01", "2");
bench_decimal_op!(sub_one_hundred, -, "2.01", "100");
bench_decimal_op!(sub_point_zero_one, -, "2.01", "0.01");
bench_decimal_op!(sub_negative_point_five, -, "2.01", "-0.5");
bench_decimal_op!(sub_pi, -, "2.01", "3.1415926535897932384626433832");
bench_decimal_op!(sub_negative_pi, -, "2.01", "-3.1415926535897932384626433832");

bench_fold_op!(sub_10k, -, 5_000_000, 10_000);

/* Mul */
bench_decimal_op!(mul_one, *, "2.01", "1");
bench_decimal_op!(mul_two, *, "2.01", "2");
bench_decimal_op!(mul_one_hundred, *, "2.01", "100");
bench_decimal_op!(mul_point_zero_one, *, "2.01", "0.01");
bench_decimal_op!(mul_negative_point_five, *, "2.01", "-0.5");
bench_decimal_op!(mul_pi, *, "2.01", "3.1415926535897932384626433832");
bench_decimal_op!(mul_negative_pi, *, "2.01", "-3.1415926535897932384626433832");

bench_fold_op!(mul_25, *, Decimal::from_str("1.1").unwrap(), 25);

/* Div */
bench_decimal_op!(div_one, /, "2.01", "1");
bench_decimal_op!(div_two, /, "2.01", "2");
bench_decimal_op!(div_one_hundred, /, "2.01", "100");
bench_decimal_op!(div_point_zero_one, /, "2.01", "0.01");
bench_decimal_op!(div_negative_point_five, /, "2.01", "-0.5");
bench_decimal_op!(div_pi, /, "2.01", "3.1415926535897932384626433832");
bench_decimal_op!(div_negative_pi, /, "2.01", "-3.1415926535897932384626433832");
bench_decimal_op!(div_no_underflow, /, "1.02343545345", "0.35454343453");
bench_fold_op!(div_10k, /, Decimal::MAX, 10_000);
bench_fold_op!(rem_10k, %, Decimal::MAX, 10_000);